engine/utils/HeapArray.h

```diff
@@ -0,0 +1,735 @@
+     */
+    void AllocateMemory(const size_t& arraySize, const size_t& bytesCount)
+    {
+        data = Allocate<T>(sizeof(T) * arraySize);
+        stateMasks = Allocate<uint8_t>(BYTE_MASK_SIZE * bytesCount);
+    }
+
+    /**
+     * @brief Copies data into the `data` pointer.
+     * @param source the pointer to copy data from.
+     * @param dataSize the amount of data to be copied.
+     */
+    void CopyData(const T* source, const size_t& dataSize)
+    {
+        memcpy(data, source, dataSize);
+    }
+
+    /**
+     * @brief Copies the byte masks from one pointer to another.
+     * @param source the pointer to copy data from.
+     * @param size the size of the memory to be copied.
+     */
+    void CopyMasks(const uint8_t* source, const size_t& byteSize)
+    {
+        memcpy(stateMasks, source, byteSize);
+    }
+
+    /**
+     * @brief Resizes the state mask in accordance to the new array size.
+     * @param newSize the new array size.
+     * @param newByteCount the new number of byte masks.
+     */
+    void ResizeStateMask(const size_t& newSize, const size_t& newByteCount)
+    {
+        // Get the last byte mask since it is the only one needed to change.
+        size_t bitsToProcess = newSize - ((newByteCount - 1) * 8);
+
+        // Get the number of bits that need to be checked for equality.
+        // i.e: if we have a mask: 0001 0101 in an array of size 5, and we resize it to only have 4
+        // elements, we want to make sure that the fifth bit in the mask is reset. As such, we want
+        // to make sure that we reset all bits EXCEPT for the first four bits.
+        size_t bitIndex = std::clamp<size_t>(bitsToProcess - 1, 0, bitsToProcess);
+
+        // Use an AND operation to reset the byte mask to the position that we want.
+        stateMasks[CalculateByteMaskCount(newSize) - 1] &= MAX_BIT_VALUES[bitIndex];
+    }
+
+    /**
+     * @brief Frees the memory allocated by the array.
+     */
+    void DeallocateMemory()
+    {
+        free(data);
+        free(stateMasks);
+    }
+
+    /**
+     * @brief Resets all counters to their default values.
+     */
+    void ResetValues()
+    {
+        data = nullptr;
+        stateMasks = nullptr;
+
+        size = count = 0;
+    }
+
+    /**
+     * @brief Copies the elements from one HeapArray to another.
+     * @param other the other array to copy elements from.
+     */
+    void Copy(const HeapArray<T>& other)
+    {
+        if (other.size == 0 && other.data == nullptr) return;
+
+        count = other.count;
+        size = other.size;
+
+        auto byteCount = CalculateByteMaskCount(other.size);
+
+        AllocateMemory(other.size, byteCount);
+
+        CopyData(other.data, sizeof(T) * size);
+        CopyMasks(other.stateMasks, BYTE_MASK_SIZE * byteCount);
+    }
+
+    // Tracker Variables
+    size_t size {0};
+    size_t count {0};
+
+    // Pointers
+    uint8_t* stateMasks {nullptr};
```

Comment:
Why does this need to be a pointer? A `uint8_t` is the same size as a pointer, so there's no space savings and it poses an additional point of error to cause a memory leak.

Reply:
Because we aren't storing a single 8-bit unsigned integer. We're storing multiple 8-bit unsigned integers for every 8 elements in the array. The pointer indicates that this is an array value.

Reply:
Ah, then rename this to something like `uint8_t* stateMaskBitfield` to be more accurate about what it is. Alternatively you could use `std::vector<bool>` since that will use a bitfield under the hood and manage it for you.

Reply:
Changed the name.